src/naive/date.rs

```diff
@@ -597,31 +597,99 @@ impl NaiveDate {
         parsed.to_naive_date()
     }
 
-    /// Private function to calculate necessary primitives for `Add<Month>`
+    /// Add a duration in [`Months`] to the date
     ///
-    /// # Arguments
+    /// If the day would be out of range for the resulting month, use the last day for that month.
     ///
-    /// * `delta` - Number of months (+/-) to add
+    /// Returns `None` if the resulting date would be out of range.
     ///
-    /// # Returns
+    /// ```
+    /// # use chrono::{NaiveDate, Months};
+    /// assert_eq!(
+    ///     NaiveDate::from_ymd(2022, 2, 20).checked_add_months(Months::new(6)),
+    ///     Some(NaiveDate::from_ymd(2022, 8, 20))
+    /// );
+    /// assert_eq!(
+    ///     NaiveDate::from_ymd(2022, 7, 31).checked_add_months(Months::new(2)),
+    ///     Some(NaiveDate::from_ymd(2022, 9, 30))
+    /// );
+    /// ```
+    pub fn checked_add_months(self, months: Months) -> Option<Self> {
+        if months.0 == 0 {
+            return Some(self);
+        }
+
+        match months.0 <= core::i32::MAX as u32 {
```

Comment:
Could we avoid this by having `try_new` and `new` constructors on the `Months` which only accept valid u32 values?